README
NAME
    File::chdir - a more sensible way to change directories

VERSION
    This documentation describes version 0.1004.

SYNOPSIS
       use File::chdir;
 
       $CWD = "/foo/bar";     # now in /foo/bar
       {
           local $CWD = "/moo/baz";  # now in /moo/baz
           ...
       }
 
       # still in /foo/bar!

DESCRIPTION
    Perl's "chdir()" has the unfortunate problem of being very, very, very
    global. If any part of your program calls "chdir()" or if any library
    you use calls "chdir()", it changes the current working directory for
    the whole program.

    This sucks.

    File::chdir gives you an alternative, $CWD and @CWD. These two variables
    combine all the power of "chdir()", File::Spec and Cwd.

$CWD
    Use the $CWD variable instead of "chdir()" and Cwd.

         use File::chdir;
         $CWD = $dir;  # just like chdir($dir)!
         print $CWD;   # prints the current working directory

    It can be localized, and it does the right thing.

         $CWD = "/foo";      # it's /foo out here.
         {
             local $CWD = "/bar";  # /bar in here
         }
         # still /foo out here!

    $CWD always returns the absolute path in the native form for the
    operating system.

    $CWD and normal "chdir()" work together just fine.

@CWD
    @CWD represents the current working directory as an array, each
    directory in the path is an element of the array. This can often make
    the directory easier to manipulate, and you don't have to fumble with
    "File::Spec->splitpath" and "File::Spec->catdir" to make portable code.

       # Similar to chdir("/usr/local/src/perl")
       @CWD = qw(usr local src perl);

    pop, push, shift, unshift and splice all work. pop and push are probably
    the most useful.

       pop @CWD;                 # same as chdir(File::Spec->updir)
       push @CWD, 'some_dir'     # same as chdir('some_dir')

    @CWD and $CWD both work fine together.

    NOTE Due to a perl bug you can't localize @CWD. See "BUGS and" for a
    work around.

EXAMPLES
    (We omit the "use File::chdir" from these examples for terseness)

    Here's $CWD instead of "chdir()":

         $CWD = 'foo';           # chdir('foo')

    and now instead of Cwd.

         print $CWD;             # use Cwd;  print Cwd::abs_path

    you can even do zsh style "cd foo bar"

         $CWD = '/usr/local/foo';
         $CWD =~ s/usr/var/;

    if you want to localize that, make sure you get the parens right

         {
             (local $CWD) =~ s/usr/var/;
             ...
         }

    It's most useful for writing polite subroutines which don't leave the
    program in some strange directory:

         sub foo {
             local $CWD = 'some/other/dir';
             ...do your work...
         }

    which is much simpler than the equivalent:

         sub foo {
             use Cwd;
             my $orig_dir = Cwd::abs_path;
             chdir('some/other/dir');
 
             ...do your work...
 
             chdir($orig_dir);
         }

    @CWD comes in handy when you want to start moving up and down the
    directory hierarchy in a cross-platform manner without having to use
    File::Spec.

         pop @CWD;                   # chdir(File::Spec->updir);
         push @CWD, 'some', 'dir'    # chdir(File::Spec->catdir(qw(some dir)));

    You can easily change your parent directory:

         # chdir from /some/dir/bar/moo to /some/dir/foo/moo
         $CWD[-2] = 'foo';

CAVEATS
   Assigning to @CWD calls "chdir()" for each element
         @CWD = qw/a b c d/;

    Internally, Perl clears @CWD and assigns each element in turn. Thus,
    this code above will do this:

         chdir 'a';
         chdir 'a/b';
         chdir 'a/b/c';
         chdir 'a/b/c/d';

    Generally, avoid assigning to @CWD and just use push and pop instead.

   "local @CWD" does not work.
    "local @CWD>" will not localize @CWD. This is a bug in Perl, you can't
    localize tied arrays. As a work around localizing $CWD will effectively
    localize @CWD.

         {
             local $CWD;
             pop @CWD;
             ...
         }

   Volumes not handled
    There is currently no way to change the current volume via File::chdir.

NOTES
    $CWD returns the current directory using native path separators, i.e. \
    on Win32. This ensures that $CWD will compare correctly with directories
    created using File::Spec. For example:

         my $working_dir = File::Spec->catdir( $CWD, "foo" );
         $CWD = $working_dir;
         doing_stuff_might_chdir();
         is( $CWD, $working_dir, "back to original working_dir?" );

    Deleting the last item of @CWD will act like a pop. Deleting from the
    middle will throw an exception.

         delete @CWD[-1]; # OK
         delete @CWD[-2]; # Dies

    What should %CWD do? Something with volumes?

         # chdir to C:\Program Files\Sierra\Half Life ?
         $CWD{C} = '\\Program Files\\Sierra\\Half Life';

DIAGNOSTICS
    If an error is encountered when changing $CWD or @CWD, one of the
    following exceptions will be thrown:

    *   *Can't delete except at the end of @CWD*

    *   *Failed to change directory to '$dir'*

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted through the web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=File-chdir>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    *   Michael G Schwern <schwern@pobox.com> (original author)

    *   David A Golden <dagolden@cpan.org> (current maintainer)

LICENSE
    Copyright 2001-2003 by Michael G Schwern <schwern@pobox.com>. Portions
    copyright 2006-2007 by David A Golden <dagolden@cpan.org>.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://dev.perl.org/licenses/>

HISTORY
    Michael wanted "local chdir" to work. p5p didn't. But it wasn't over!
    Was it over when the Germans bombed Pearl Harbor? Hell, no!

    Abigail and/or Bryan Warnock suggested the $CWD thing (Michael forgets
    which). They were right.

    The "chdir()" override was eliminated in 0.04.

    David became co-maintainer with 0.06_01 to fix some chronic Win32 path
    bugs.

    As of 0.08, if changing $CWD or @CWD fails to change the directory, an
    error will be thrown.

SEE ALSO
    File::pushd, File::Spec, Cwd, "chdir" in perlfunc, "Animal House"
    <http://www.imdb.com/title/tt0077975/quotes>



Changes
Changes for File-chdir

{{ $NEXT }}

0.1011    2016-10-07 10:23:15-04:00 America/New_York

    [Documentation]

    - Fixed POD typos

0.1010    2015-02-05 10:04:27-05:00 America/New_York

    [Changed]

    - Updated distribution metadata and repo layout

    - Changed bugtracker to github

0.1009    2014-09-23 12:24:11-04:00 America/New_York

    [Changed]

    - Updated distribution metadata

0.1008    2012-12-02 22:32:37 America/New_York

    - Fixed broken use of abs_path on Cygwin [Joel Berger]

0.1007    2012-09-17 20:29:43 America/New_York

    - Fixed broken Pod link; reordered CAVEATS section

0.1006    2011-11-02 18:11:13 America/New_York

    - Handle directories with newlines [rt.cpan.org #72114]

0.1005    2011-11-02 14:21:27 America/New_York

    - Fixed rt.cpan.org #53064 [Michael Schwern]

    - Removes bundled Test::More

    - Converted distribution builder to Dist::Zilla

0.1004 Wed Jun  9 14:14:54 EDT 2010

    - Build.Pl and Makefile.PL will bail out on Perl 5.13.1, which
      accidentally broke File::chdir

0.1003 Tue Apr 27 21:21:32 EDT 2010

    - testfix: remove -T flag in tests

0.1002 Mon Feb 11 19:47:46 EST 2008

    - testfix: 0.1001 accidentally left many array tests commented out

0.1001 Sat Feb  9 13:07:05 EST 2008

    - testfix: work around odd Test::Builder errors on perl 5.6.2 dying in
      the middle of a "push"

0.10 Fri Feb  8 08:42:53 EST 2008
    - File::Spec 3.2501 broke File::chdir on Win32; updated prerequisite to
      latest File::Spec (3.27) and rewrote internals of File::chdir
    - Split array deletion tests to separate file to skip on Perl < 5.006
    - Added Build.PL and other files (e.g. INSTALL, LICENSE) to the tarball
    - Minor documentation changes/additions

0.0901
    - removed unnecessary prototypes on internal functions (perl critic)
    - cosmetic changes to distribution package (README, extra tests in xt)

0.09 Mon Jul 30 09:50:28 EDT 2007
    - Updated Makefile.PL to warn about API change in 0.08
    - Added $! diagnostic to error messages

0.08 Thu Jun 21 20:56:50 EDT 2007
    * Croaks if the chdir fails when changing $CWD or @CWD
    * Deleting from end of @CWD acts like a pop
    - Revised and expanded @CWD tests
    - Revised delete error message for deleting from middle of @CWD
    - Added DIAGNOSTICS section to documentation

0.07 Mon Jun 18 18:38:00 EDT 2007
    - Fixed testfile bug that prepended "/" in tests on Cygwin

0.06_01 Mon Apr  9 05:17:02 EDT 2007
    - Development release by David Golden
    * Fixed bug that prepended "/" to $CWD for Win32
    * $CWD changed to use native path separators
    * Tests fixed on Win32 by always using native separators as well
    * Upped File::Spec and Cwd dependencies to require many recent bug-fixes
    - Added note about comparing $CWD to File::Spec generated names
    - Other minor documentation patches from Schwern

0.06  Thu Aug 14 17:02:32 PDT 2003
    * Now working under taint mode (thanks Mark Guckeyson)
    - Small nit in the SYNOPSIS
    * Removed dependency on File::Spec 0.8 and File::Spec::Functions
    - Added a LICENSE
    - Added NOTES about %CWD.

0.05  Mon Nov 26 15:44:20 EST 2001
    * Added @CWD
    - $CWD's value now guaranteed to be correct even if some other code
      calls chdir().

0.04  Mon Nov 26 13:36:24 EST 2001  *UNRELEASED*
    * The magic chdir() is *gone*
    * Now works back to 5.004

0.03  Mon Sep  3 20:10:15 EDT 2001
    * chdir() is being deprecated and will be removed next version!
    - Gave credit to Abigail and Bryan

0.02  Mon Sep  3 19:51:02 EDT 2001
    * Added $CWD
    - chdir() with no args works

0.01  Fri Aug 31 21:15:25 EDT 2001
    * HA!  It works!



Build.PL
use 5.004;
use strict;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";

# Do Schwern's API warnings and make sure it appends to generated Makefile.PL
my $api_warning = do { local $/; <DATA> };
eval $api_warning;
if ( $@ ) {
  die $@ if $@ =~ /OS unsupported/;
  warn $@;
}

$api_warning = quotemeta($api_warning);

my $mb = $class->subclass( code => << "END_CODE" );

    sub do_create_makefile_pl {
        my \$self = shift;
        warn "Custom Makefile.PL\n";
        \$self->SUPER::do_create_makefile_pl;
        local *FH;
        open FH, ">> Makefile.PL" or die \$!;
        print {FH} "\nBEGIN {\n$api_warning\n}\n";
        close FH;
    }

END_CODE

$mb->new(
    module_name         => 'File::chdir',
    dist_author         => 'David A. Golden <dagolden@cpan.org>',
    license             => 'perl',
    create_readme       => 1,
    create_makefile_pl  => 'traditional',
    requires        => {
        # module requirements here
        'Cwd'           => 3.16,
        'File::Spec'    => 3.27,
    },
    meta_add            => {
        no_index => {
            directory => [ qw{
                examples
                inc
            }],
        }
    },
)->create_build_script;

__DATA__
die "OS unsupported: Perl 5.13.1 accidentally broke File::chdir. Aborting!\n"
  if $] eq "5.013001";

use vars qw/$PACKAGE $LAST_API_CHANGE/;
$PACKAGE = 'File::chdir';
$LAST_API_CHANGE = 0.08;

eval "require $PACKAGE";

no strict 'refs';
if(!$@ && ${$PACKAGE.'::VERSION'} < $LAST_API_CHANGE) {
    print <<"CHANGE_WARN";

WARNING  WARNING  WARNING  WARNING  WARNING  WARNING  WARNING  WARNING

-----*****>>>>>     THE INTERFACE HAS CHANGED     <<<<<*****----------

As of 0.08, changing \$CWD or \@CWD will die with an error if the directory
change fails.  Earlier versions of File::chdir failed silently.

WARNING  WARNING  WARNING  WARNING  WARNING  WARNING  WARNING  WARNING

We'll pause for 20 seconds...

CHANGE_WARN

    sleep 20;
}




Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "a more sensible way to change directories",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>, Michael G. Schwern <schwern\@pobox.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "File-chdir",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "File::chdir",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Cwd" => "3.16",
    "Exporter" => 0,
    "File::Spec::Functions" => "3.27",
    "strict" => 0,
    "vars" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.1012",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Cwd" => "3.16",
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => "3.27",
  "Test::More" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/File/chdir.pm
package File::chdir;
use 5.004;
use strict;
use vars qw($VERSION @ISA @EXPORT $CWD @CWD);
# ABSTRACT: a more sensible way to change directories

our $VERSION = '0.1012';

require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(*CWD);

use Carp;
use Cwd 3.16;
use File::Spec::Functions 3.27 qw/canonpath splitpath catpath splitdir catdir/;

tie $CWD, 'File::chdir::SCALAR' or die "Can't tie \$CWD";
tie @CWD, 'File::chdir::ARRAY'  or die "Can't tie \@CWD";

sub _abs_path {
    # Otherwise we'll never work under taint mode.
    my($cwd) = Cwd::getcwd =~ /(.*)/s;
    # Run through File::Spec, since everything else uses it
    return canonpath($cwd);
}

# splitpath but also split directory
sub _split_cwd {
    my ($vol, $dir) = splitpath(_abs_path, 1);
    my @dirs = splitdir( $dir );
    shift @dirs; # get rid of leading empty "root" directory
    return ($vol, @dirs);
}

# catpath, but take list of directories
# restore the empty root dir and provide an empty file to avoid warnings
sub _catpath {
    my ($vol, @dirs) = @_;
    return catpath($vol, catdir(q{}, @dirs), q{});
}

sub _chdir {
    # Untaint target directory
    my ($new_dir) = $_[0] =~ /(.*)/s;

    local $Carp::CarpLevel = $Carp::CarpLevel + 1;
    if ( ! CORE::chdir($new_dir) ) {
        croak "Failed to change directory to '$new_dir': $!";
    };
    return 1;
}

{
    package File::chdir::SCALAR;
    use Carp;

    BEGIN {
        *_abs_path = \&File::chdir::_abs_path;
        *_chdir = \&File::chdir::_chdir;
        *_split_cwd = \&File::chdir::_split_cwd;
        *_catpath = \&File::chdir::_catpath;
    }

    sub TIESCALAR {
        bless [], $_[0];
    }

    # To be safe, in case someone chdir'd out from under us, we always
    # check the Cwd explicitly.
    sub FETCH {
        return _abs_path;
    }

    sub STORE {
        return unless defined $_[1];
        _chdir($_[1]);
    }
}


{
    package File::chdir::ARRAY;
    use Carp;

    BEGIN {
        *_abs_path = \&File::chdir::_abs_path;
        *_chdir = \&File::chdir::_chdir;
        *_split_cwd = \&File::chdir::_split_cwd;
        *_catpath = \&File::chdir::_catpath;
    }

    sub TIEARRAY {
        bless {}, $_[0];
    }

    sub FETCH {
        my($self, $idx) = @_;
        my ($vol, @cwd) = _split_cwd;
        return $cwd[$idx];
    }

    sub STORE {
        my($self, $idx, $val) = @_;

        my ($vol, @cwd) = _split_cwd;
        if( $self->{Cleared} ) {
            @cwd = ();
            $self->{Cleared} = 0;
        }

        $cwd[$idx] = $val;
        my $dir = _catpath($vol,@cwd);

        _chdir($dir);
        return $cwd[$idx];
    }

    sub FETCHSIZE {
        my ($vol, @cwd) = _split_cwd;
        return scalar @cwd;
    }
    sub STORESIZE {}

    sub PUSH {
        my($self) = shift;

        my $dir = _catpath(_split_cwd, @_);
        _chdir($dir);
        return $self->FETCHSIZE;
    }

    sub POP {
        my($self) = shift;

        my ($vol, @cwd) = _split_cwd;
        my $popped = pop @cwd;
        my $dir = _catpath($vol,@cwd);
        _chdir($dir);
        return $popped;
    }

    sub SHIFT {
        my($self) = shift;

        my ($vol, @cwd) = _split_cwd;
        my $shifted = shift @cwd;
        my $dir = _catpath($vol,@cwd);
        _chdir($dir);
        return $shifted;
    }

    sub UNSHIFT {
        my($self) = shift;

        my ($vol, @cwd) = _split_cwd;
        my $dir = _catpath($vol, @_, @cwd);
        _chdir($dir);
        return $self->FETCHSIZE;
    }

    sub CLEAR  {
        my($self) = shift;
        $self->{Cleared} = 1;
    }

    sub SPLICE {
        my $self = shift;
        my $offset = shift || 0;
        my $len = shift || $self->FETCHSIZE - $offset;
        my @new_dirs = @_;

        my ($vol, @cwd) = _split_cwd;
        my @orig_dirs = splice @cwd, $offset, $len, @new_dirs;
        my $dir = _catpath($vol, @cwd);
        _chdir($dir);
        return @orig_dirs;
    }

    sub EXTEND { }
    sub EXISTS {
        my($self, $idx) = @_;
        return $self->FETCHSIZE >= $idx ? 1 : 0;
    }

    sub DELETE {
        my($self, $idx) = @_;
        croak "Can't delete except at the end of \@CWD"
            if $idx < $self->FETCHSIZE - 1;
        local $Carp::CarpLevel = $Carp::CarpLevel + 1;
        $self->POP;
    }
}

1;
__END__

=head1 SYNOPSIS

  use File::chdir;

  $CWD = "/foo/bar";     # now in /foo/bar
  {
      local $CWD = "/moo/baz";  # now in /moo/baz
      ...
  }

  # still in /foo/bar!

=head1 DESCRIPTION

Perl's C<chdir()> has the unfortunate problem of being very, very, very
global.  If any part of your program calls C<chdir()> or if any library
you use calls C<chdir()>, it changes the current working directory for
the *whole* program.

This sucks.

File::chdir gives you an alternative, C<$CWD> and C<@CWD>.  These two
variables combine all the power of C<chdir()>, L<File::Spec> and L<Cwd>.

=head1 $CWD

Use the C<$CWD> variable instead of C<chdir()> and Cwd.

    use File::chdir;
    $CWD = $dir;  # just like chdir($dir)!
    print $CWD;   # prints the current working directory

It can be localized, and it does the right thing.

    $CWD = "/foo";      # it's /foo out here.
    {
        local $CWD = "/bar";  # /bar in here
    }
    # still /foo out here!

C<$CWD> always returns the absolute path in the native form for the
operating system.

C<$CWD> and normal C<chdir()> work together just fine.

=head1 @CWD

C<@CWD> represents the current working directory as an array, each
directory in the path is an element of the array.  This can often make
the directory easier to manipulate, and you don't have to fumble with
C<< File::Spec->splitpath >> and C<< File::Spec->catdir >> to make portable code.

  # Similar to chdir("/usr/local/src/perl")
  @CWD = qw(usr local src perl);

pop, push, shift, unshift and splice all work.  pop and push are
probably the most useful.

  pop @CWD;                 # same as chdir(File::Spec->updir)
  push @CWD, 'some_dir'     # same as chdir('some_dir')

C<@CWD> and C<$CWD> both work fine together.

*NOTE* Due to a perl bug you can't localize C<@CWD>.  See L</CAVEATS> for a work around.

=head1 EXAMPLES

(We omit the C<use File::chdir> from these examples for terseness)

Here's C<$CWD> instead of C<chdir()>:

    $CWD = 'foo';           # chdir('foo')

and now instead of Cwd.

    print $CWD;             # use Cwd;  print Cwd::abs_path

you can even do zsh style C<cd foo bar>

    $CWD = '/usr/local/foo';
    $CWD =~ s/usr/var/;

if you want to localize that, make sure you get the parens right

    {
        (local $CWD) =~ s/usr/var/;
        ...
    }

It's most useful for writing polite subroutines which don't leave the
program in some strange directory:

    sub foo {
        local $CWD = 'some/other/dir';
        ...do your work...
    }

which is much simpler than the equivalent:

    sub foo {
        use Cwd;
        my $orig_dir = Cwd::getcwd;
        chdir('some/other/dir');

        ...do your work...

        chdir($orig_dir);
    }

C<@CWD> comes in handy when you want to start moving up and down the
directory hierarchy in a cross-platform manner without having to use
File::Spec.

    pop @CWD;                   # chdir(File::Spec->updir);
    push @CWD, 'some', 'dir'    # chdir(File::Spec->catdir(qw(some dir)));

You can easily change your parent directory:

    # chdir from /some/dir/bar/moo to /some/dir/foo/moo
    $CWD[-2] = 'foo';

=head1 CAVEATS

=head2 C<local @CWD> does not work.

C<local @CWD> will not localize C<@CWD>.  This is a bug in Perl, you
can't localize tied arrays.  As a work around localizing $CWD will
effectively localize @CWD.

    {
        local $CWD;
        pop @CWD;
        ...
    }

=head2 Assigning to C<@CWD> calls C<chdir()> for each element

    @CWD = qw/a b c d/;

Internally, Perl clears C<@CWD> and assigns each element in turn.  Thus, this
code above will do this:

    chdir 'a';
    chdir 'a/b';
    chdir 'a/b/c';
    chdir 'a/b/c/d';

Generally, avoid assigning to C<@CWD> and just use push and pop instead.

=head2 Volumes not handled

There is currently no way to change the current volume via File::chdir.

=head1 NOTES

C<$CWD> returns the current directory using native path separators, i.e. C<\>
on Win32.  This ensures that C<$CWD> will compare correctly with directories
created using File::Spec.  For example:

    my $working_dir = File::Spec->catdir( $CWD, "foo" );
    $CWD = $working_dir;
    doing_stuff_might_chdir();
    is( $CWD, $working_dir, "back to original working_dir?" );

Deleting the last item of C<@CWD> will act like a pop.  Deleting from the
middle will throw an exception.

    delete @CWD[-1]; # OK
    delete @CWD[-2]; # Dies

What should %CWD do?  Something with volumes?

    # chdir to C:\Program Files\Sierra\Half Life ?
    $CWD{C} = '\\Program Files\\Sierra\\Half Life';

=head1 DIAGNOSTICS

If an error is encountered when changing C<$CWD> or C<@CWD>, one of
the following exceptions will be thrown:

* ~Can't delete except at the end of @CWD~
* ~Failed to change directory to '$dir'~

=head1 HISTORY

Michael wanted C<local chdir> to work.  p5p didn't.  But it wasn't over!
Was it over when the Germans bombed Pearl Harbor?  Hell, no!

Abigail and/or Bryan Warnock suggested the C<$CWD> thing (Michael forgets
which).  They were right.

The C<chdir()> override was eliminated in 0.04.

David became co-maintainer with 0.06_01 to fix some chronic
Win32 path bugs.

As of 0.08, if changing C<$CWD> or C<@CWD> fails to change the directory, an
error will be thrown.

=head1 SEE ALSO

L<File::pushd>, L<File::Spec>, L<Cwd>, L<perlfunc/chdir>,
"Animal House" L<http://www.imdb.com/title/tt0077975/quotes>

=cut

